prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,188 @@
+
+from prompt2model.utils.config import DEFAULT_HYPERPARAMETERS, DEFAULT_HYPERPARAMETERS_SPACE
+from prompt2model.model_trainer import BaseTrainer
+from prompt2model.param_selector.base import ParamSelector
+
+# TODO:
+# - User tweaking hyperparameter
+# - Dynamic initialization of hyperparamter range from task type and complexity
+# - Using LLM to suggest hyperparameter
+
+
+class OptunaParamSelector(ParamSelector):
+    """Uses optuna for searching for hyperparameters"""
+
+    def __init__(self, trainer: BaseTrainer):
+        """Initialize with train/val datasets and a prompt specification"""
+        self.trainer = trainer
+    
+    @property
+    def _example_hyperparameters(self) -> dict[str, Any]:
+        """Example hyperparameters (for testing only)."""
+        return DEFAULT_HYPERPARAMETERS
+    
+    @property
+    def _example_hyperparameter_space(self) -> dict[str, Any]:
+        return DEFAULT_HYPERPARAMETERS_SPACE
+
+    def select_from_hyperparameters(
+        self,
+        training_sets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: dict[str, Any],
+    ) -> transformers.Trainer:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_sets (list[Dataset]): One or more training datasets for the trainer.
+            validation (Dataset): A dataset for computing validation metrics.
+            hyperparameters (dict[str, list[Any]]): A dictionary of hyperparameter choices.
+
+        Supported keys for the hyperparameters and their specs. Left side we have the key and
+        have the expected type of the value in braces.
+            - :key min_num_train_epochs: (int)
+            - :key max_num_train_epochs: (int)
+            - :key save_strategy: List[str]
+            - :evaluation_strategy: List[str], available options: ["epoch", "no"]
+            - :per_device_train_batch_size: List[int], available options: ["epoch", "steps", "no"],
+            - :min_weight_decay: float
+            - :max_weight_decay: float
+            - :min_learning_rate: float
+            - :max_learning_rate: float
+        Here is a example of hyperparameters:
+        ```python
+        hyperparameters = {
+            "min_num_train_epochs": 5,
+            "max_num_train_epochs": 10,
+            "save_strategy": ["epoch", "steps", "no"],
+            "evaluation_strategy": ["epoch", "no"],
+            "per_device_train_batch_size": [4, 8, 16, 32],
+            "min_weight_decay": 1e-5,
+            "max_weight_decay": 1e-1,
+            "min_learning_rate": 1e-5,
+            "max_learning_rate", 1e-1,
+        }
+        ```
+        Returns:
+            Returns the transformers.Trainer class with the optimal hyperparameters
+        """
+        # TODO:
+        # - Find the industry standards for default values spec
+        # - More asserts for other keys. Example checking the min or max values
```

Comment:
Delete the TODOs and add as issues. Also, we/others can help choose these default values over time.